src/java/htsjdk/samtools/sra/SRALazyRecord.java

```diff
@@ -0,0 +1,941 @@
+*               National Center for Biotechnology Information
+*
+*  This software/database is a "United States Government Work" under the
+*  terms of the United States Copyright Act.  It was written as part of
+*  the author's official duties as a United States Government employee and
+*  thus cannot be copyrighted.  This software/database is freely available
+*  to the public for use. The National Library of Medicine and the U.S.
+*  Government have not placed any restriction on its use or reproduction.
+*
+*  Although all reasonable efforts have been taken to ensure the accuracy
+*  and reliability of the software and data, the NLM and the U.S.
+*  Government do not and cannot warrant the performance or results that
+*  may be obtained by using this software or data. The NLM and the U.S.
+*  Government disclaim all warranties, express or implied, including
+*  warranties of performance, merchantability or fitness for any particular
+*  purpose.
+*
+*  Please cite the author in any work or product based on this material.
+*
+* ===========================================================================
+*
+*/
+
+package htsjdk.samtools.sra;
+
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.SAMTagUtil;
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.Cigar;
+import htsjdk.samtools.SAMBinaryTagAndValue;
+import htsjdk.samtools.SAMValidationError;
+import htsjdk.samtools.util.Log;
+import ngs.ReadCollection;
+import ngs.AlignmentIterator;
+import ngs.Alignment;
+import ngs.ReadIterator;
+import ngs.Read;
+import ngs.Fragment;
+import ngs.ErrorMsg;
+
+import java.util.Set;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.List;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+/**
+ * Created by andrii.nikitiuk on 8/25/15.
+ */
+public class SRALazyRecord extends SAMRecord {
+    private static final Log log = Log.getInstance(SRALazyRecord.class);
+
+    private boolean isAligned;
+    private ReadCollection run;
+    private Object sraIterator;
+    private String sraReadId;
+    private String sraAlignmentId;
+    private int unalignedReadFragmentIndex = -1;
+
+
+    private Set<LazyField> initializedFields = new HashSet<LazyField>();
+    private Set<LazyFlag> initializedFlags = new HashSet<LazyFlag>();
+    private Set<LazyAttribute> initializedAttributes = new HashSet<LazyAttribute>();
+
+    private enum LazyField {
+        ALIGNMENT_START,
+        MAPPING_QUALITY,
+        REFERENCE_NAME,
+        CIGAR_STRING,
+        BASES,
+        QUALS,
+        MATE_ALIGNMENT_START,
+        MATE_REFERENCE_NAME,
+        INFERRED_INSERT_SIZE
+    }
+
+    private enum LazyFlag {
+        READ_NEGATIVE_STRAND,
+        READ_PAIRED,
+        NOT_PRIMARY_ALIGNMENT,
+        MATE_NEGATIVE_STRAND,
+        MATE_UNMAPPED,
+        FIRST_OF_PAIR,
+        SECOND_OF_PAIR
+    }
+
+    private enum LazyAttribute {
+        RG
```

Comment:
SRA format does not store SAM attributes, but rather store other data that makes it possible to compute original attributes values. But since GATK3 just calls SAMRecord.getBinaryAttributes() and then uses only some of them, it does not make sense to calculate their values which might be not used later. What is more, all of sam attributes are optional and sam/bam files are not required to have them. Concluding all said, SRA will only support RG attribute for now and we can discuss if there are any other attributes which are required for GATK4.